File: cMsg/java/org/jlab/coda/cMsg/apps/cMsgMonitor.java
package org.jlab.coda.cMsg.apps;

import org.jlab.coda.cMsg.*;

/**
 * Application that gets and prints monitor data.
 */
public class cMsgMonitor {

    String  name = "monitor";
    String  description = "java monitor";
    String  UDL = "cMsg:cMsg://aslan:3456/cMsg/test";
    boolean debug;


    /** Constructor. */
    cMsgMonitor(String[] args) {
        decodeCommandLine(args);
    }


    /**
     * Method to decode the command line used to start this application.
     * @param args command line arguments
     */
    public void decodeCommandLine(String[] args) {

        // loop over all args
        for (int i = 0; i < args.length; i++) {

            if (args[i].equalsIgnoreCase("-h")) {
                usage();
                System.exit(-1);
            }
            else if (args[i].equalsIgnoreCase("-n")) {
                name = args[i + 1];
                i++;
            }
            else if (args[i].equalsIgnoreCase("-d")) {
                description = args[i + 1];
                i++;
            }
            else if (args[i].equalsIgnoreCase("-u")) {
                UDL= args[i + 1];
                i++;
            }
            else if (args[i].equalsIgnoreCase("-debug")) {
                debug = true;
            }
            else {
                usage();
                System.exit(-1);
            }
        }

        return;
    }


    /** Method to print out correct program command line usage. */
    private static void usage() {
        System.out.println("\nUsage:\n\n" +
            "   java cMsgConsumer [-n name] [-d description] [-u UDL] [-debug]\n");
    }


    /**
     * Run as a stand-alone application.
     */
    public static void main(String[] args) {
        try {
            cMsgMonitor mon = new cMsgMonitor(args);
            mon.run();
        }
        catch (cMsgException e) {
            e.printStackTrace();
            System.exit(-1);
        }
    }


    /**
     * This method is executed as a thread.
     */
    public void run() throws cMsgException {

        if (debug) {
            System.out.println("Running cMsg monitor");
        }

        // connect to cMsg server
        cMsg coda = new cMsg(UDL, name, description);
        coda.connect();

        long period = 3000; // millisec

        cMsgMessage msg;

        while (true) {
            msg = coda.monitor(null);
            System.out.println("Msg ->\n" + msg.getText());
            // wait
            try { Thread.sleep(period); }
            catch (InterruptedException e) {}
        }
    }
}

File: cMsg/src/execsrc/cMsgMonitor.cc
//  cMsgMonitor.cc
//
//  Monitors cMsg system
//
//  E.Wolin, 17-may-2007



// system includes
using namespace std;
#include <iostream>
#include <unistd.h>


// for cMsg
#include <cMsg.hxx>
using namespace cmsg;


// connection parameters
static string udl;
static string name;
static string description;
static int repeat_time = 3;


// prototypes
void decodeCommandLine(int argc, char **argv);



//-----------------------------------------------------------------------------


int main(int argc, char **argv) {


  // set defaults
  udl           = "cMsg://broadcast/cMsg";
  name          = "cMsgMonitor";
  description   = "cMsgMonitor Utility ";


  // decode command line parameters
  decodeCommandLine(argc,argv);


  // create cMsg object
  cMsg *c;
  try {
    c = new cMsg(udl,name,description);
  } catch (exception &e) {
    cerr << e.what() << endl;
    exit(EXIT_FAILURE);
  } catch (...) {
    cerr << "?unknown exception" << endl;
    exit(EXIT_FAILURE);
  }
  

  // connect and loop forever
  try {
    c->connect();

    while(true) {
      cMsgMessage *m = c->monitor("");
      cout << m->getText() << endl;
      cout << "----------------------------------------------------------" << endl << endl;
      delete(m);
      sleep(repeat_time);
    }
    
  } catch (exception &e) {
    cerr << e.what() << endl;
    exit(EXIT_FAILURE);
  } catch (...) {
    cerr << "?unknown exception" << endl;
    exit(EXIT_FAILURE);
  }
  
  exit(EXIT_SUCCESS);
}


//-----------------------------------------------------------------------------


void decodeCommandLine(int argc, char **argv) {
  

  const char *help = 
    "\nusage:\n\n   cMsgMonitor [-udl udl] [-n name] [-d description] [-r repeat_time]\n\n";
  
  

  // loop over arguments
  int i=1;
  while (i<argc) {
    if (strncasecmp(argv[i],"-h",2)==0) {
      cout << help << endl;
      exit(EXIT_SUCCESS);

    } else if (strncasecmp(argv[i],"-udl",4)==0) {
      udl=argv[i+1];
      i=i+2;

    } else if (strncasecmp(argv[i],"-n",2)==0) {
      name=argv[i+1];
      i=i+2;

    } else if (strncasecmp(argv[i],"-d",2)==0) {
      description=argv[i+1];
      i=i+2;
    } else if (strncasecmp(argv[i],"-r",2)==0) {
      repeat_time=atoi(argv[i+1]);
      i=i+2;
    }
  }

}


//-----------------------------------------------------------------------------
